reconcile/utils/aws_api_typed/api.py

```diff
@@ -0,0 +1,194 @@
+from __future__ import annotations
+
+import textwrap
+from abc import ABC, abstractmethod
+from typing import Any, TypeVar
+
+from boto3 import Session
+from botocore.client import BaseClient
+from pydantic import BaseModel
+
+import reconcile.utils.aws_api_typed.iam
+import reconcile.utils.aws_api_typed.organization
+import reconcile.utils.aws_api_typed.sts
+from reconcile.utils.aws_api_typed.iam import AWSApiIam
+from reconcile.utils.aws_api_typed.organization import AWSApiOrganizations
+from reconcile.utils.aws_api_typed.sts import AWSApiSts
+
+SubApi = TypeVar("SubApi")
+
+
+class AWSCredentials(ABC):
+    @abstractmethod
+    def as_env_vars(self) -> dict[str, str]:
+        """
+        Returns a dictionary of environment variables that can be used to authenticate with AWS.
+        """
+        ...
+
+    @abstractmethod
+    def as_credentials_file(self, profile_name: str = "default") -> str:
+        """
+        Returns a string that can be used to write an AWS credentials file.
+        """
+        ...
+
+    @abstractmethod
+    def build_session(self) -> Session:
+        """
+        Builds an AWS session using these credentials.
+        """
+        ...
+
+    def get_temporary_credentials(
+        self, aws_api: AWSApi, duration_seconds: int = 900
+    ) -> AWSTemporaryCredentials:
+        """
+        Builds temporary AWS credentials from a session. This is similar to assuming a role,
+        in the sense that the credentials will expire after a certain amount of time.
+
+        These temporary credentials have the same permissions as the session they were built from, except:
+        - they can't be used for anything IAM related
+        - for the STS API only the AssumeRole and GetSessionToken actions are allowed
+        """
+        tmp_creds = aws_api.sts.get_session_token(duration_seconds=duration_seconds)
+        return AWSTemporaryCredentials(
+            access_key_id=tmp_creds.access_key_id,
+            secret_access_key=tmp_creds.secret_access_key,
+            session_token=tmp_creds.session_token,
+            region=aws_api.session.region_name,
+        )
```

Comment:
how about you move this function to `AWSApi`.  after the refactor this function does not really belong here anymore, wdyt?

Reply:
Good idea; applied in d4ff0c8